File: lokma/geometry/strategies.py
"""Mass-estimation strategies (Strategy pattern).

  * :class:`PixelRatioStrategy` — the Phase 1 2D heuristic (``portion_g · area/ref_area``).
  * :class:`VolumetricStrategy` — ``m = V·ρ`` from the per-frame :class:`ScaleEstimate`
    + a per-class height prior.
  * :class:`AutoStrategy` — **the default**: volumetric when calibration confidence
    clears ``config.calibration_confidence_threshold``, else pixel_ratio tagged
    ``"uncalibrated"``. This is the safe-activation gate.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from dataclasses import dataclass, field

from lokma.config import AppConfig
from lokma.core.exceptions import ConfigurationError
from lokma.core.models import (
    MAX_FORESHORTENING_TILT_DEG,
    Detection,
    FoodRecord,
    MassEstimate,
    ScaleEstimate,
)
from lokma.density.categories import height_for, porosity_for
from lokma.density.density_service import DensityService
from lokma.geometry.depth_volume_engine import DepthVolumeEngineService
from lokma.geometry.volume_engine import VolumeEngineService


@dataclass
class MassContext:
    """Per-frame dependencies for a strategy. ``scale`` is recomputed each frame."""

    scale: ScaleEstimate | None
    volume_engine: VolumeEngineService
    density_service: DensityService
    config: AppConfig
    depth_engine: DepthVolumeEngineService = field(default_factory=DepthVolumeEngineService)


class MassEstimationStrategy(ABC):
    name: str = "base"

    @abstractmethod
    def estimate(self, detection: Detection, food: FoodRecord, ctx: MassContext) -> MassEstimate:
        ...


class PixelRatioStrategy(MassEstimationStrategy):
    """2D heuristic: scale a known portion by the ratio of observed to reference area."""

    name = "pixel_ratio"

    def estimate(self, detection, food, ctx):
        ref_area = food.ref_area if (food.ref_area and food.ref_area > 0) else ctx.config.default_ref_area
        scale = detection.mask_area_px / ref_area if ref_area > 0 else 0.0
        grams = (food.portion_g or 0.0) * scale
        return MassEstimate(grams=grams, method=self.name, density_used=None)


def _effective_porosity(detection, food) -> float:
    """Effective porosity P for ``× (1 − P)``: the per-instance ML-predicted value when
    present, else the per-class category fallback (mirrors the DensityService hierarchy)."""
    if detection.predicted_porosity is not None:
        return detection.predicted_porosity
    return porosity_for(food.class_name)


class VolumetricStrategy(MassEstimationStrategy):
    """Physical path: metric footprint area → volume → mass via ``m = V·ρ·(1−P)``."""

    name = "volumetric"

    def estimate(self, detection, food, ctx):
        if ctx.scale is None:
            raise ConfigurationError("VolumetricStrategy requires a ScaleEstimate")
        # Native-frame area (square pixels) keeps the metric scaling correct.
        area_px = detection.mask_area_px_frame or detection.mask_area_px
        real_area_cm2 = ctx.scale.area_px_to_cm2(area_px)
        density, density_source = ctx.density_service.resolve(food)
        shape = food.geometric_shape or "prism"
        height_cm = height_for(food.class_name)
        volume = ctx.volume_engine.estimate_volume(real_area_cm2, shape, height_cm)
        grams = volume.volume_cm3 * density * (1.0 - _effective_porosity(detection, food))
        return MassEstimate(
            grams=grams,
            method=f"{self.name}:{density_source.value}",
            density_used=density,
            volume_cm3=volume.volume_cm3,
            calibration_source=ctx.scale.source.value,
            calibration_confidence=ctx.scale.confidence,
        )


def _tilt_trustworthy(tilt_deg: float | None) -> bool:
    """Volumetric area scaling degrades past a grazing incidence; gate it off there."""
    return tilt_deg is None or tilt_deg <= MAX_FORESHORTENING_TILT_DEG


class AutoStrategy(MassEstimationStrategy):
    """Safe activation: volumetric only when calibration is confident enough."""

    name = "auto"

    def __init__(self) -> None:
        self._volumetric = VolumetricStrategy()
        self._pixel_ratio = PixelRatioStrategy()

    def estimate(self, detection, food, ctx):
        # 1. Measured LiDAR depth volume — most accurate, and inherently tilt-robust
        #    (the fitted plane absorbs orientation), so it needs no tilt gate.
        if detection.depth_sample is not None:
            dv = ctx.depth_engine.integrate_sample(detection.depth_sample)
            if dv is not None:
                density, density_source = ctx.density_service.resolve(food)
                return MassEstimate(
                    grams=dv.volume_cm3 * density * (1.0 - _effective_porosity(detection, food)),
                    method=f"volumetric_depth:{density_source.value}",
                    density_used=density,
                    volume_cm3=dv.volume_cm3,
                    calibration_source="depth_plane",
                    calibration_confidence=dv.coverage,
                )
        # 2. Scalar volumetric when calibration is confident and not too oblique.
        scale = ctx.scale
        threshold = ctx.config.calibration_confidence_threshold
        if scale is not None and scale.confidence >= threshold and _tilt_trustworthy(scale.tilt_deg):
            return self._volumetric.estimate(detection, food, ctx)
        # 3. Not trustworthy enough (low confidence or too oblique) — fall back and mark it.
        base = self._pixel_ratio.estimate(detection, food, ctx)
        return MassEstimate(
            grams=base.grams,
            method=f"{base.method}:uncalibrated",
            density_used=base.density_used,
            volume_cm3=base.volume_cm3,
            calibration_source=(scale.source.value if scale else None),
            calibration_confidence=(scale.confidence if scale else None),
        )


_STRATEGIES: dict[str, type[MassEstimationStrategy]] = {
    PixelRatioStrategy.name: PixelRatioStrategy,
    VolumetricStrategy.name: VolumetricStrategy,
    AutoStrategy.name: AutoStrategy,
}


def get_strategy(name: str) -> MassEstimationStrategy:
    """Instantiate a strategy by name (``config.strategy``)."""
    try:
        return _STRATEGIES[name]()
    except KeyError as exc:
        raise ConfigurationError(
            f"Unknown strategy '{name}'. Available: {sorted(_STRATEGIES)}"
        ) from exc

File: ios/LokmaCore/Sources/LokmaCore/Strategies.swift
// Mass-estimation strategies — mirror of `lokma/geometry/strategies.py`.
//
//  * PixelRatioStrategy — 2D heuristic (portion · area/ref_area).
//  * VolumetricStrategy — m = V·ρ from the per-frame ScaleEstimate + height prior.
//  * AutoStrategy — the default: volumetric when calibration confidence clears
//    the threshold, else pixel_ratio tagged ":uncalibrated".
import Foundation

public enum StrategyError: Error {
    case missingScale
    case unknownStrategy(String)
}

/// Per-frame dependencies for a strategy. `scale` is recomputed each frame.
public struct MassContext {
    public let scale: ScaleEstimate?
    public let volumeEngine: VolumeEngine
    public let densityService: DensityService
    public let depthEngine: DepthVolumeEngine
    public let config: AppConfig

    public init(scale: ScaleEstimate?, volumeEngine: VolumeEngine = VolumeEngine(),
                densityService: DensityService = DensityService(),
                depthEngine: DepthVolumeEngine = DepthVolumeEngine(), config: AppConfig = AppConfig()) {
        self.scale = scale
        self.volumeEngine = volumeEngine
        self.densityService = densityService
        self.depthEngine = depthEngine
        self.config = config
    }
}

public protocol MassEstimationStrategy {
    var name: String { get }
    func estimate(_ detection: Detection, _ food: FoodRecord, _ ctx: MassContext) throws -> MassEstimate
}

public struct PixelRatioStrategy: MassEstimationStrategy {
    public let name = "pixel_ratio"
    public init() {}

    public func estimate(_ detection: Detection, _ food: FoodRecord, _ ctx: MassContext) throws -> MassEstimate {
        let refArea = (food.refArea != nil && food.refArea! > 0) ? food.refArea! : ctx.config.defaultRefArea
        let scale = refArea > 0 ? detection.maskAreaPx / refArea : 0.0
        let grams = food.portionG * scale
        return MassEstimate(grams: grams, method: name, densityUsed: nil)
    }
}

public struct VolumetricStrategy: MassEstimationStrategy {
    public let name = "volumetric"
    public init() {}

    public func estimate(_ detection: Detection, _ food: FoodRecord, _ ctx: MassContext) throws -> MassEstimate {
        guard let scale = ctx.scale else { throw StrategyError.missingScale }
        // Native-frame area (square pixels) keeps the metric scaling correct;
        // 0.0 is falsy in Python, so fall back to the working-grid area.
        let areaPx = detection.maskAreaPxFrame != 0 ? detection.maskAreaPxFrame : detection.maskAreaPx
        let realAreaCm2 = scale.areaPxToCm2(areaPx)
        let shape = food.geometricShape ?? "prism"
        let heightCm = Categories.heightFor(food.className)
        let volume = ctx.volumeEngine.estimateVolume(realAreaCm2: realAreaCm2, shape: shape, heightCm: heightCm)
        let m = volumetricGrams(detection, food, volume.volumeCm3, ctx)
        return MassEstimate(
            grams: m.grams,
            method: "\(name):\(m.methodTag)",
            densityUsed: m.densityUsed,
            volumeCm3: volume.volumeCm3,
            calibrationSource: scale.source.rawValue,
            calibrationConfidence: scale.confidence
        )
    }
}

public struct AutoStrategy: MassEstimationStrategy {
    public let name = "auto"
    private let volumetric = VolumetricStrategy()
    private let pixelRatio = PixelRatioStrategy()
    public init() {}

    /// Volumetric area scaling degrades past a grazing incidence; gate it off there.
    static func tiltTrustworthy(_ tiltDeg: Double?) -> Bool {
        guard let t = tiltDeg else { return true }
        return t <= ScaleEstimate.maxForeshorteningTiltDeg
    }

    public func estimate(_ detection: Detection, _ food: FoodRecord, _ ctx: MassContext) throws -> MassEstimate {
        // 1. Measured LiDAR depth volume — most accurate, and inherently tilt-robust
        //    (the fitted plane absorbs orientation), so it needs no tilt gate.
        if let sample = detection.depthSample, let dv = ctx.depthEngine.integrate(sample) {
            let m = volumetricGrams(detection, food, dv.volumeCm3, ctx)
            return MassEstimate(
                grams: m.grams,
                method: "volumetric_depth:\(m.methodTag)",
                densityUsed: m.densityUsed,
                volumeCm3: dv.volumeCm3,
                calibrationSource: "depth_plane",
                calibrationConfidence: dv.coverage
            )
        }
        // 2. Scalar volumetric when calibration is confident and not too oblique.
        let scale = ctx.scale
        let threshold = ctx.config.calibrationConfidenceThreshold
        if let scale, scale.confidence >= threshold, Self.tiltTrustworthy(scale.tiltDeg) {
            return try volumetric.estimate(detection, food, ctx)
        }
        // 3. Not trustworthy enough (low confidence or too oblique) — fall back and mark it.
        let base = try pixelRatio.estimate(detection, food, ctx)
        return MassEstimate(
            grams: base.grams,
            method: "\(base.method):uncalibrated",
            densityUsed: base.densityUsed,
            volumeCm3: base.volumeCm3,
            calibrationSource: scale?.source.rawValue,
            calibrationConfidence: scale?.confidence
        )
    }
}

/// Effective porosity P for the mass correction `× (1 − P)`: the per-instance
/// ML-predicted value when present, else the per-class category fallback (mirrors the
/// DensityService resolution hierarchy). `Mass = V · ρ · (1 − P)`.
public func effectivePorosity(_ detection: Detection, _ food: FoodRecord) -> Double {
    detection.predictedPorosity ?? Categories.porosityFor(food.className)
}

/// Convert a volume (cm³) to grams for the volumetric paths. Prefers the Phase-8 ML
/// fill-density head (`mass = V·D`, where D already folds ρ and porosity) when present,
/// else the analytic `mass = V·ρ·(1−P)`. The global calibration constant (default 1.0)
/// corrects the absolute V→mass scale for the device's real intrinsics.
public func volumetricGrams(_ detection: Detection, _ food: FoodRecord, _ volumeCm3: Double,
                            _ ctx: MassContext) -> (grams: Double, densityUsed: Double, methodTag: String) {
    let k = ctx.config.massCalibrationConstant
    if let d = detection.predictedFillDensity {
        return (k * volumeCm3 * d, d, "fill")
    }
    let (density, source) = ctx.densityService.resolve(food)
    let grams = k * volumeCm3 * density * (1.0 - effectivePorosity(detection, food))
    return (grams, density, source.rawValue)
}

public func makeStrategy(_ name: String) throws -> MassEstimationStrategy {
    switch name {
    case "pixel_ratio": return PixelRatioStrategy()
    case "volumetric": return VolumetricStrategy()
    case "auto": return AutoStrategy()
    default: throw StrategyError.unknownStrategy(name)
    }
}
